operator/pkg/upgrade/upgrade.go

```diff
@@ -2,411 +2,248 @@ package upgrade
-	if err != nil {
-		return fmt.Errorf("ensure artifacts job for nodes: %w", err)
+	// if there is a difference between what we want and what we have
+	// we should update the cluster instead of letting chart errors stop deployment permanently
+	// otherwise if there are errors we need to abort
+	if len(chartErrors) > 0 && !chartDrift {
+		chartErrorString := strings.Join(chartErrors, ",")
+		chartErrorString = "failed to update helm charts: " + chartErrorString
+		fmt.Printf("Chart errors: %s\n", chartErrorString)
+		return fmt.Errorf("helm charts have errors and there is no update to be applied")
 	}
 
-	log.Info("Waiting for artifacts to be placed on nodes...")
-
-	err = wait.PollUntilContextCancel(ctx, 5*time.Second, true, func(ctx context.Context) (bool, error) {
-		jobs, err := artifacts.ListArtifactsJobForNodes(ctx, cli, in)
-		if err != nil {
-			return false, fmt.Errorf("list artifacts jobs for nodes: %w", err)
-		}
+	// If all addons match their target version + values, things are successful
+	// This should not happen on upgrades
+	if len(pendingCharts) == 0 && !chartDrift {
+		return nil
+	}
 
-		ready := true
-		for nodeName, job := range jobs {
-			if job == nil {
-				return false, fmt.Errorf("job for node %s not found", nodeName)
-			}
-			if job.Status.Succeeded > 0 {
-				continue
-			}
-			ready = false
-			for _, cond := range job.Status.Conditions {
-				if cond.Type == batchv1.JobFailed {
-					if cond.Status == corev1.ConditionTrue {
-						// fail immediately if any job fails
-						return false, fmt.Errorf("job for node %s failed: %s - %s", nodeName, cond.Reason, cond.Message)
-					}
-					break
-				}
-			}
-			// job is still running
-		}
+	if len(pendingCharts) > 0 {
+		// If there are pending charts, return an error because we need to wait for some prior installation to complete
+		return fmt.Errorf("pending charts: %v", pendingCharts)
+	}
 
-		return ready, nil
-	})
-	if err != nil {
-		return fmt.Errorf("wait for artifacts job for nodes: %w", err)
+	if !chartDrift {
+		// if there is no drift, we should not reapply the cluster config
+		// This should not happen on upgrades
+		return nil
 	}
 
-	log.Info("Artifacts placed on nodes")
+	// Replace the current chart configs with the new chart configs
+	clusterConfig.Spec.Extensions.Helm = cfgs
+	fmt.Printf("Updating cluster config with new helm charts %v\n", changedCharts)
+	//Update the clusterConfig
+	if err := cli.Update(ctx, &clusterConfig); err != nil {
+		return fmt.Errorf("failed to update cluster config: %w", err)
+	}
 	return nil
-}
-
-func ensureAirgapArtifactsInCluster(ctx context.Context, cli client.Client, in *clusterv1beta1.Installation) error {
-	log := ctrl.LoggerFrom(ctx)
 
-	log.Info("Uploading container images...")
+}
 
-	err := autopilotEnsureAirgapArtifactsPlan(ctx, cli, in)
+func unLockInstallation(ctx context.Context, cli client.Client, in *v1beta1.Installation) error {
```

Comment:
can you explain to me and comment what this does? why do you copy the new spec into the old installation? i thought we create new installation crs for upgrades?

Reply:
yeah this definitely needs a comment then
this isn't the 'old' installation, this is the installation created at the beginning of the upgrade process - it's what we're upgrading to
but because that was created before the CRD was updated, it might not have all the fields, and so updating it with the from-configmap version is needed

Reply:
Actually I dont follow. The installation used to be created here

https://github.com/replicatedhq/embedded-cluster/pull/1222/files#diff-1595f9a51283933adb1a95f0d4a42f0576e783a1932c0d0da27fa1e5ca5628b3L69

```
	err = createInstallation(ctx, cli, in)
	if err != nil {
		return fmt.Errorf("apply installation: %w", err)
	}
```